x-pack/plugins/features/server/feature_schema.ts

```diff
@@ -61,73 +99,190 @@ const schema = Joi.object({
+
+  function validateManagementEntry(
+    privilegeId: string,
+    managementEntry: Record<string, string[]> = {}
+  ) {
+    Object.entries(managementEntry).forEach(([managementSectionId, managementSectionEntry]) => {
+      if (featureManagement.has(managementSectionId)) {
+        managementSectionEntry.forEach(entry => {
+          featureManagement.get(managementSectionId)!.delete(entry);
+          if (featureManagement.get(managementSectionId)?.size === 0) {
+            featureManagement.delete(managementSectionId);
+          }
+        });
+      }
+      if (!management[managementSectionId]) {
+        throw new Error(
+          `Feature privilege ${feature.id}.${privilegeId} has unknown management section: ${managementSectionId}`
+        );
+      }
+
+      const unknownSectionEntries = difference(
+        managementSectionEntry,
+        management[managementSectionId]
+      );
+
+      if (unknownSectionEntries.length > 0) {
+        throw new Error(
+          `Feature privilege ${
+            feature.id
+          }.${privilegeId} has unknown management entries for section ${managementSectionId}: ${unknownSectionEntries.join(
+            ', '
+          )}`
+        );
+      }
+    });
+  }
+
+  const privilegeEntries: Array<[string, FeatureKibanaPrivileges]> = [];
+  if (feature.privileges !== 'none') {
+    privilegeEntries.push(...Object.entries(feature.privileges));
+  }
+  if (feature.reserved) {
+    privilegeEntries.push(['reserved', feature.reserved.privilege]);
+  }
+
+  if (privilegeEntries.length === 0) {
+    return;
+  }
+
+  privilegeEntries.forEach(([privilegeId, privilegeDefinition]) => {
+    if (!privilegeDefinition) {
+      throw new Error('Privilege definition may not be null or undefined');
+    }
 
-    Object.entries(privilegeDefinition.management || {}).forEach(
-      ([managementSectionId, managementEntry]) => {
-        if (!management[managementSectionId]) {
-          throw new Error(
-            `Feature privilege ${feature.id}.${privilegeId} has unknown management section: ${managementSectionId}`
-          );
-        }
+    validateAppEntry(privilegeId, privilegeDefinition.app);
+
+    validateCatalogueEntry(privilegeId, privilegeDefinition.catalogue);
+
+    validateManagementEntry(privilegeId, privilegeDefinition.management);
+  });
 
-        const unknownSectionEntries = difference(managementEntry, management[managementSectionId]);
+  // Seed this list with the "fixed" set of primary/minimal feature privileges.
+  // This initial set is fixed because the primary feature privileges are currently required to
+  // be "all" and "read". Security, when enabled, will add "minimal" versions of these two privileges
+  // in order to support sub-feature privileges.
+  // This set should be reevaluated If/when the primary feature privileges support flexible definitions beyond "all" and "read".
+  const seenPrivilegeIds: Set<string> = new Set(['all', 'read', 'minimal_all', 'minimal_read']);
```

Comment:
How do you feel about moving this validation to within the security plugin, since it knows about the serialization format for these privileges? This would potentially let us be even lazier (and more fool proof) by serializing all of the privileges to their ES format and ensuring there aren't duplicates.

Reply:
I can see what this would look like. I like that we have most/all of the validation in one place now, and splitting it out could make it harder to follow.


I'm not sure if it makes a huge difference (it likely doesn't), but moving this check to the security plugin also means that the validation won't happen until the `start` phase. Today, we fail this at the `setup` phase, which causes the plugin registering the feature to fail `setup`. Moving this logic to the security plugin will cause the `security` plugin to fail `start`, even though it hasn't done anything wrong. The offending plugin would still be able to start successfully.

Reply:
> I can see what this would look like. I like that we have most/all of the validation in one place now, and splitting it out could make it harder to follow.

Yeah... having all of the validation in one place is convenient. I also started trying to think through how we could change the privilege serialization format to prevent conflicts, but all of the solutions I came up with impact the Role CRUD APIs, and likely require significant changes there.

Do you think it's worthwhile to try this out and see how you feel about it?

Reply:
Turns out it's not that bad; I'll push changes once you're finished with the first round of reviews, I don't want to clobber your current mental model :)